### PR TITLE
Make courses only accessible by their audiences

### PR DESCRIPTION
**Pull request by dmcavoy**
Finally locking courses down to the instructor and participant audiences they are set to!

This makes it so only those that are in the audiences for the course will be able to view UnitGroup, Script, Lesson, and ScriptLevel pages for that course. On `show` we will display the existing no_access page with error message. For all other actions it will show forbidden. 

(Apologies about the very ugly commit history. This PR has been in flight for a long time)

## Links

Redirection Logic was updated in https://github.com/code-dot-org/code-dot-org/pull/44097

[Eng Plan](https://docs.google.com/document/d/1V61xONU0-mleMEEdHWNxhvZtYIPRCB31mT_hbUXusCQ/edit#)
[Product Spec](https://docs.google.com/document/d/1qZpc90daxQiiqpcQUiUiGV2tQoaYEQUzOA72UZzBlVE/edit#heading=h.f5x8no3lbpxy)

## Testing story

- Will be bug bashed before releasing
- Added a ton of new unit tests (Most of the lines changed in this PR are new tests)

### CODE REVIEW

File: dashboard/test/models/ability_test.rb [hunk 1]
```diff
@@ -1,50 +1,123 @@
-require 'test_helper'
-
 class AbilityTest < ActiveSupport::TestCase
```
Comment (dmcavoy) [inline dashboard/test/models/ability_test.rb]: Add tests for pilot unit and pilot pl unit

File: dashboard/app/models/script.rb [hunk 2]
```diff
@@ -601,12 +601,20 @@ def self.get_unit_family_redirect_for_user(family_name, user: nil, locale: 'en-U
     family_units = Script.get_family_from_cache(family_name).sort_by(&:version_year).reverse
 
     # Only students should be redirected based on unit progress and/or section assignments.
+    # TODO(dmcavoy): should only participants redirected?
     if user&.student?
```
Comment (dmcavoy) [inline dashboard/app/models/script.rb]: Do we want participants or students to be redirected?

File: dashboard/app/models/concerns/curriculum/course_types.rb [hunk 3]
```diff
@@ -53,6 +53,8 @@ def can_be_participant?(user)
     # If unit is in a unit group then decide based on unit group audience
     return unit_group.can_be_participant?(user) if is_a?(Script) && unit_group
 
+    # Signed out users can only use student facing courses
+    return false if !user && participant_audience != 'student'
```
Comment (dmcavoy) [inline dashboard/app/models/concerns/curriculum/course_types.rb]: Make its own PR

**Comment by dmcavoy** [comment]
Does this redirect people to sign in if they try to access something where participant audience is not student?

https://github.com/code-dot-org/code-dot-org/pull/44101#discussion_r770937181

File: dashboard/app/models/unit_group.rb [hunk 4]
```diff
@@ -530,15 +530,18 @@ def redirect_to_course_url(user)
   # @param user [User]
   # @return [Boolean] Whether the user can view the course.
   def can_view_version?(user = nil)
+    # Must be part of instructor or participant audience to view version
+    return false unless can_be_participant?(user) || can_be_instructor?(user)
```
Comment (davidsbailey) [inline dashboard/app/models/unit_group.rb]: this reminds me that it seems worth testing what happens on family name redirect for users who "cannot be participant" of any scripts/courses in the family name

**Comment by dmcavoy** [comment]
As people review I would be interested to hear if people feel like we need the adhoc to have a clone of the production database to test this or not. I'm leaning toward no seeing where the changes ended up

File: dashboard/app/controllers/courses_controller.rb [hunk 5]
```diff
@@ -173,6 +173,11 @@ def check_plc_enrollment
   end
 
   def render_no_access
+    unless @unit_group.can_be_instructor?(current_user) || @unit_group.can_be_participant?(current_user)
+      authenticate_user!
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/courses_controller.rb]: I'm not familiar with `authenticate_user` but why would we want to authenticate the user if we're going to return :no_access anyway?

File: dashboard/app/models/ability.rb [hunk 6]
```diff
@@ -261,48 +261,48 @@ def initialize(user)
...
     can [:read, :show_by_id, :student_lesson_plan], Lesson do |lesson|
       script = lesson.script
-      if script.in_development?
-        user.permission?(UserPermission::LEVELBUILDER)
-      elsif script.pilot?
-        script.has_pilot_access?(user)
-      else
-        true
-      end
+      (script.can_be_participant?(user) || script.can_be_instructor?(user)) && can?(:read, script)
```
Comment (jamescodeorg) [inline dashboard/app/models/ability.rb]: Do we need to check for `script.can_be_participant?` and `script.can_be_instructor?` here or is `can?(:read, script)` by itself sufficient?  (same question for script_level)

File: dashboard/app/models/unit_group.rb [hunk 7]
```diff
@@ -530,15 +530,18 @@ def redirect_to_course_url(user)
...
+
```
Comment (jamescodeorg) [inline dashboard/app/models/unit_group.rb]: I'm a little fuzzy on when and where we use CanCan.  Would this be a good place to use CanCan?  (Similar comment for `script`).

File: dashboard/app/controllers/courses_controller.rb [hunk 8]
```diff
@@ -173,6 +173,11 @@ def check_plc_enrollment
   end
 
   def render_no_access
+    unless @unit_group.can_be_instructor?(current_user) || @unit_group.can_be_participant?(current_user)
```
Comment (jamescodeorg) [inline dashboard/app/controllers/courses_controller.rb]: Would it make sense to replace the check here and in scripts_controller.rb with CanCan?

File: dashboard/app/controllers/courses_controller.rb (hunk 5 above)
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb]: `authenticate_user` makes sure to redirect a signed out user to the sign in page but I have refactored to deal with that.

File: dashboard/app/models/ability.rb (hunk 6 above)
Comment (dmcavoy) [reply dashboard/app/models/ability.rb]: Thanks for the push on this. I have updated!

File: dashboard/app/models/unit_group.rb (hunk 7 above)
Comment (dmcavoy) [reply dashboard/app/models/unit_group.rb]: Updated!

File: dashboard/app/controllers/courses_controller.rb (hunk 8 above)
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb]: Updated!

File: dashboard/app/models/script.rb [hunk 9]
```diff
@@ -737,7 +739,7 @@ def can_view_version?(user, locale: nil)
 
     # Restrictions only apply to students and logged out users.
```
Comment (dmcavoy) [inline dashboard/app/models/script.rb]: Update this comment to say participants

**Review by jamescodeorg — APPROVED** [review]
LGTM, thanks for all the new tests!

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/app/models/script.rb [hunk 10]
```diff
@@ -705,6 +725,8 @@ def link
   # @param locale [String] User or request locale. Optional.
   # @return [Boolean] Whether the user can view the unit.
   def can_view_version?(user, locale: nil)
+    return false unless Ability.new(user).can?(:read, self)
```
Comment (davidsbailey) [inline dashboard/app/models/script.rb:728]: it's great to see authorization logic being consolidated into CanCan!

File: dashboard/test/controllers/lessons_controller_test.rb [hunk 11]
```diff
@@ -97,6 +97,27 @@ class LessonsControllerTest < ActionController::TestCase
...
+      script_id: @pilot_pl_script.id,
+      lesson_group: pilot_pl_lesson_group,
+      name: 'Pilot Lesson 1',
+      absolute_position: 1,
+      relative_position: 1,
+      has_lesson_plan: true,
+      lockable: false,
+      properties: {
+        overview: 'lesson overview',
+        student_overview: 'student overview'
```
Comment (davidsbailey) [inline dashboard/test/controllers/lessons_controller_test.rb]: is there anything special about this lesson? if not, can you try generating it using `create :script, :with_lessons` instead? https://github.com/code-dot-org/code-dot-org/blob/fe459f188996dc07056a524ae1891be4219d2aa4/dashboard/test/factories/factories.rb#L761

if you need to add or modify anything after you create it, you can always do that after: https://github.com/code-dot-org/code-dot-org/blob/9ceee45455abeb12601b619d3e102fa780a06024/dashboard/test/controllers/lessons_controller_test.rb#L574-L577

this should make it easier to see if you are just creating a lesson or if there is anything special about it.

File: dashboard/test/controllers/lessons_controller_test.rb [hunk 12]
```diff
@@ -113,6 +134,36 @@ class LessonsControllerTest < ActionController::TestCase
         student_overview: 'student overview'
       }
     )
+
+    @pl_login_req_script = create :script, name: 'signed-in-pl-script', is_migrated: true, include_student_lesson_plans: true, login_required: true, instructor_audience: SharedCourseConstants::INSTRUCTOR_AUDIENCE.facilitator, participant_audience: SharedCourseConstants::PARTICIPANT_AUDIENCE.teacher
+    pl_login_req_lesson_group = create :lesson_group, script: @pl_login_req_script
+    @pl_login_req_lesson = create(
```
Comment (davidsbailey) [inline dashboard/test/controllers/lessons_controller_test.rb]: can you use `:with_lessons` instead here too?

File: dashboard/test/controllers/courses_controller_test.rb [hunk 13]
```diff
@@ -16,12 +16,23 @@ class CoursesControllerTest < ActionController::TestCase
...
 
     @migrated_unit = create :script, is_migrated: true, published_state: SharedCourseConstants::PUBLISHED_STATE.beta
     @unit_group_migrated = create :unit_group, published_state: SharedCourseConstants::PUBLISHED_STATE.beta
     create :unit_group_unit, unit_group: @unit_group_migrated, script: @migrated_unit, position: 1
 
+    @migrated_pl_unit = create :script, is_migrated: true, published_state: SharedCourseConstants::PUBLISHED_STATE.beta
+    @pl_unit_group_migrated = create :unit_group, published_state: SharedCourseConstants::PUBLISHED_STATE.beta, instructor_audience: SharedCourseConstants::INSTRUCTOR_AUDIENCE.facilitator, participant_audience: SharedCourseConstants::PARTICIPANT_AUDIENCE.teacher
+    create :unit_group_unit, unit_group: @pl_unit_group_migrated, script: @migrated_pl_unit, position: 1
+    @migrated_pl_unit.reload
+
```
Comment (davidsbailey) [inline dashboard/test/controllers/courses_controller_test.rb:29]: can you please move all the lines you've added to `setup`, plus lines 14-17, to a new `setup_all` block? it looks like this will shave 30-60 seconds off the time it takes to run this test.

this test was already quite slow to begin with, so if there's anything else you can move into `setup_all` without breaking tests, that would be a nice bonus. as a guideline, if the object is not being modified within individual tests then it is a good candidate to extract to `setup_all`.

**Review by davidsbailey — APPROVED** [review]
great test coverage Dani! I think most of my comments around testing could be addressed in a follow-up PR.

File: dashboard/app/models/unit_group.rb (hunk 4 above)
Comment (dmcavoy) [reply dashboard/app/models/unit_group.rb]: Added

**Comment by dmcavoy** [comment]
@davidsbailey @bethanyaconnor @jamescodeorg Want to take one final look? I fixed the issues that came up during the [bug bash](https://docs.google.com/document/d/1PUj72r9p8oig9IGHVhz9PWbBYHnIGYICPjUXjg_PbaY/edit#) with viewing levels you don't have access to.

Screen Shot 2022-02-22 at 11 42 00 AM

File: dashboard/test/factories/factories.rb [hunk 14]
```diff
@@ -766,7 +766,7 @@
       after(:create) do |script, evaluator|
         lesson_group = create :lesson_group, script: script
         evaluator.lessons_count.times do
-          create :lesson, :with_activity_section, lesson_group: lesson_group
+          create :lesson, :with_activity_section, lesson_group: lesson_group, script: script, has_lesson_plan: true
```
Comment (davidsbailey) [inline dashboard/test/factories/factories.rb:769]: good find!

**Review by davidsbailey — APPROVED** [review]
woohoo! 🎉
